Add support for changing essential boundary conditions

rcarson3 opened the pull request:
This will close #8 and #27.

This PR adds a fairly major new feature to the code that'll be quite useful to many different users in the form of actually now being able to simulate cyclic loading, strain rate jump tests, and more.

I also added a nice life of change feature of allowing the average stress txt file to be named whatever the user wants. One nice benefit of this is it also allows us to run our python test suite in parallel. So, all of those tests don't take so long.

Before this PR goes in, I'll need to push an update to our MFEM branch as well. The changes there are breaking for the develop branch, so I've not pushed them up yet. However, they are necessary in order to add changing boundary conditions in a sane manner with out having to constantly destroy and recreate one object.

rcarson3 committed:
Added local action for the gradient mult term
Changing boundary conditions will require the use of the local mult operator for the gradient term. Since, it's pretty much the same as the pre-existing Mult function minus the ess bcs we now have Mult and LocalMult based on a templated version of the functions.

rcarson3 committed:
More modifications to BCManager to allow for eventual changing essential BCs

rcarson3 committed:
Framework needed for changing boundary conditions but not compiled yet

rcarson3 committed:
Working changing BCs framework
A simple test case still needs to be added to test this for something other than with the initial time step

rcarson3 committed:
Update test suite results for 1st time step
The new initial time step formulation changes the 1st time step stress ever so minorly but these values still needed to be updated or else the tests failed.

rcarson3 committed:
Merge branch 'exaconstit-dev' into changing_ess_bcs

rcarson3 committed:
Add support to option file for changing BCs

rcarson3 committed:
Add a cyclic test inorder to test out changing BCs

rcarson3 commented:
I've now added support for changing boundary conditions to the option file. I've also added a simple cyclic example to test it out.

I can add further tests where the cyclic example is all purely elastic, and a simple example where we essentially do the monotonic tension example just to show that it works as  expected. I will mention that I have done these already, and obtained the  answers that I expected.

rcarson3 committed:
Code clean-up to remove code no longer needed

rcarson3 committed:
Allow average stress file name to be set in options toml file

rcarson3 committed:
Update options.toml to include descriptions for all the new features

rcarson3 committed:
Update python test driver to run in parallel
This required some hacking around with the cmake file and dealing with the unittest framework... I'm pretty sure there will be a better way to do this. However, it currently severly cuts down on the run time of tests if you have more than 2 processors.

rcarson3 committed:
Update README

rcarson3 commented:
It would probably be useful for this PR to also update the macroscopic stress strain plotting python script and have it just go ahead and slurp up the option toml file and auto-generate the appropriate curves for you.

rcarson3 commented:
Also I think once mfem/mfem#1842 is implemented, we can probably improve on this some more, and potentially allow users to have normals of a face constrained in a given direction while the body is deforming.

nrbarton commented on src/options.toml:
```diff
@@ -34,6 +34,43 @@
 #All of these options are required
 #If they are not provided the program will exit early
 [BCs]
+    # Tells the program that we'll have changing BCs
```
In line 4 above "#Also while the below is indented to make things easier to read the parser doesn't care." ... I am not sure what you mean. Does not care about comment lines? About indentation?

rcarson3 replied:
So, the line was referring to the fact that the TOML format doesn't care about indentations. I'll make that line clearer in a follow-up commit.

nrbarton commented on src/options.toml:
```diff
@@ -34,6 +34,43 @@
 [BCs]
+    # Tells the program that we'll have changing BCs
+    # If left out of the file this defaults to a value of false
+    changing_ess_bcs = false
```
In this whole file, it might be easier to read if the comments were more clearly associated with the entries to which the comments pertain. 

Perhaps more like:
```
  # this setting does fun stuff
  foo = 1.0

  # this setting is optional and defaults to false
  bar = false
```

rcarson3 replied:
I imagine that should be a simple enough fix.

nrbarton commented on src/options.toml:
```diff
@@ -34,6 +34,43 @@
+    # You'll want to have all the BC ids that you plan on using through out the
+    # simulation listed here. Also, when constraining the problem you'll want to
+    # take into consideration that currently we can't constrain a boundary to have
+    # a desired normal through out the deformation history.
```
You mean that you cannot do mixed boundary conditions in which only some of the components are treated as essential BCs? I am not sure what you mean by "to have a desired normal." Maybe you are talking about a boundary remaining planar but allowing in-plane motion (so "mixed" as I indicate above)? But it looks like mixed conditions are allowed. So you are talking about allowing a boundary to both translate in the direction of its normal *and* remain planar?

Also -- "through out" should be "throughout"

rcarson3 replied:
So, we can do general essential boundary conditions in that you can constrain any of the dofs. However, we don't allow a user to say that they want the normal of a given boundary/plane to remain a given direction through-out deformation if a velocity is not supplied for that boundary. So, I mean the latter part of your statement.

nrbarton replied:
That is what I figured after poking around a bit more. You might consider revising the comment to clarify what you mean.

nrbarton commented on src/options.toml:
```diff
@@ -34,6 +34,43 @@
+    # take into consideration that currently we can't constrain a boundary to have
+    # a desired normal through out the deformation history.
+    #essential BC ids for the whole boundary
+    #essential_ids = [[1, 2, 3, 4],
+    #                 [1, 2, 3, 4],
+    #                 [1, 2, 3, 4],
+    #                 [1, 2, 3, 4],
+    #                 [1, 2, 3, 4]]
```
So you need a list for each entry in update_steps? And they have to be the same list each time? The user should really have to repeat the entries if they are required to be the same throughout? 

Or the ID listings can change length -- you can drop an essential BC at some update_steps entry?

rcarson3 replied:
A user does need to list the attributes for each entry in updated_steps. The reason for this is that the user can change the applied boundary condition set to whatever they want. For example, they could go from being symmetrical BCs to essential BCs at a given step. So, we don't require them to be the same through out the simulation.

nrbarton commented on src/options.toml:
```diff
@@ -34,6 +34,43 @@
+    #                   [3, 1, 2, 3]]
+    #Vector of vals to be applied for each attribute
+    #The length of this should be #ids * dim of problem
+    #essential_vals = [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.001],
+    #                  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -0.001],
+    #                  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.001],
+    #                  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -0.001],
+    #                  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.001]]
```
So if these are velocity values then you do not hold a truly steady velocity gradient. Would it make sense, as part of the changing-essential-BCs feature improvement to allow for a case in which you specify a velocity gradient and it uses the coordinates to compute the velocity for each time step? That might be a feature that folks would appreciate.

rcarson3 replied:
So, I would say prescribing a constant velocity gradient through out the deformation of the simulation is not super simple outside of maybe uniaxial loading where it simplifies to something like having a constant strain rate. I might be able to support that uniaxial case, but I would probably want to do that after this PR. Since, it'll take me a little bit to make sure we're doing it right.

I imagine once we add support for periodic boundary conditions then we should also be able to add support for a constant velocity gradient to be applied.

nrbarton replied:
A simple implementation ought to be not much more than v_i = L_ij x_j with x being the current position of the given node. Or perhaps v=L.(x-x0) if you want to generalize so that the user need not have a corner of the domain at x=0. But it can certainly wait for a future PR.

rcarson3 replied:
Yeah that might work assuming it's only done on the BCs and takes into account what dofs are constrained. I'll need to figure out a nice way to have it implemented, so we get out a v_i that corresponds to time: (t_i+1 + t_i) / 2.

nrbarton commented on src/options.toml:
```diff
@@ -126,11 +165,11 @@
     ref_par = 0
-    #The polynomial refinement/order of our shape functions
+    #The polynomial order of our shape functions
     prefinement = 1
```
This might be easier to read if it were `p_refinement` rather than `prefinement`

rcarson3 replied:
Yeah, I'd agree with that being a good change.

nrbarton commented on src/options.toml:
```diff
@@ -34,6 +34,43 @@
+    changing_ess_bcs = false
+    # The below gives an example of how to make use of changing essential BCs
+    # where we have the sample cyclically loaded.
+    # It would go in tandem with using the auto time stepper with a dt = 0.1
```
The "auto" time stepping means fixed dt? Maybe `fixed` is then a better name? I think of auto time stepping as adjusting the time step size as the simulation progresses. Which can be a really useful feature ...

rcarson3 replied:
It does mean fixed dt. I can change the name for the case of a future auto time stepping scheme. However, I'll want to do a deprecation time period and give people time to swap over to the new name. A future auto-time stepping feature will also probably require me to modify the changing BC feature to allow the user to specify a time to change the BCs at.

nrbarton commented on src/options.toml:
```diff
@@ -34,6 +34,43 @@
+    # If left out of the file this defaults to a value of false
+    changing_ess_bcs = false
+    # The below gives an example of how to make use of changing essential BCs
+    # where we have the sample cyclically loaded.
```
Cyclic loading might be easier to specify if you add a feature where the specified velocity can be a function of time (say through a 1D table), rather than making users spell things out in this rather verbose way. 

That way of doing it might dovetail nicely with a capability for constant velocity gradient type BCs.

rcarson3 replied:
I believe I should be able to add a cyclic or uniaxial type option. Since, we know the applied boundary faces remain constant through out time. However, it should be up to the user to specify the step number that things are changing and what the new velocity should be. I say this because cyclic examples can get pretty interesting where you might change the strain things are being cycled at or potentially the strain rates for which it is occurring. Alternatively, they might want to unload slightly at a given point in-order to mimic an experiment such as a HEXD cyclic exp where it is common to see the sample unloaded slightly when imagining is being taken, so you don't potentially have the sample creep.

I can add this feature in a new PR that's smaller in scope.

nrbarton reviewed (commented):
I poke at the options.toml file, as that seemed like the part on which you were most interested in feedback.

barker29 commented:
If you have any input on the interface etc in mfem/mfem#1842 I would appreciate any thoughts you have. That PR should be ready for review soon, and it will be easier to tweak the interface before it's merged.

rcarson3 commented:
> If you have any input on the interface etc in [mfem/mfem#1842](https://github.com/mfem/mfem/pull/1842) I would appreciate any thoughts you have. That PR should be ready for review soon, and it will be easier to tweak the interface before it's merged.

@barker29  I'll try and look over your PR in more detail tomorrow. Although, I might be interested in having a chat maybe over like a webex to understand how everything in there works from a bit of a higher lever sometime in the future. Since, I have more of a mechanics background then an applied math background, so I can't say I have the firmest grasp for how some of these constraints might need to be setup.

rcarson3 committed:
Address PR comments
This includes updating the reference options.toml, renaming options to things that make more sense, and then updating all the test files for these names

rcarson3 committed:
Final small changes for PR
